### PR TITLE
refactor: use `BlockWithSenders` in executors

### LINKED ISSUE #5767: Improve Execution trait ergonomics
### Describe the feature

currently the Executor trait takes the block and the recovered senders as different arguments:

https://github.com/paradigmxyz/reth/blob/a8e2518aa0f4b14e79e8f90391686402f231a450/crates/storage/provider/src/traits/executor.rs#L33-L38

this is easy to mess up, resulting in additional sender recovery.

Ideally, these functions should accept a single argument that can have multiple variants.

I think this type should be either a generic or an enum over the block variants.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by onbjerg**
Removes the optional `senders` parameter in the executors and requires passing a `BlockWithSenders` instead. Various utility functions exist on the block types to seal/unseal, provide senders and recover senders from signatures already. Since the `senders` parameter already required the user to possibly handle recovering themselves, we explicitly require it now, since it makes it easier to figure out how often you recover, which is an expensive operation.

See the refd issue for more info

Closes https://github.com/paradigmxyz/reth/issues/5767

### CODE REVIEW

File: bin/reth/src/debug_cmd/in_memory_merkle.rs [hunk 1]
```diff
@@ -166,9 +166,9 @@ impl Command {
         let merkle_block_td =
             provider.header_td_by_number(merkle_block_number)?.unwrap_or_default();
         executor.execute_and_verify_receipt(
-            &block.clone().unseal(),
+            //todo
+            &block.clone().unseal().with_recovered_senders().unwrap(),
```
Comment (onbjerg) [inline bin/reth/src/debug_cmd/in_memory_merkle.rs:170]: Not sure what error to return here

File: crates/storage/provider/src/traits/executor.rs [hunk 2]
```diff
@@ -32,19 +32,17 @@ pub trait BlockExecutor {
     /// recovery costs.
     fn execute(
```
Comment (mattsse) [inline crates/storage/provider/src/traits/executor.rs:33]: the docs need to be updated, no senders argument

File: crates/revm/src/optimism/processor.rs [hunk 3]
```diff
@@ -84,7 +79,7 @@ impl<'a> BlockExecutor for EVMProcessor<'a> {
 
         let mut cumulative_gas_used = 0;
         let mut receipts = Vec::with_capacity(block.body.len());
-        for (transaction, sender) in block.body.iter().zip(senders) {
+        for (transaction, sender) in block.body.iter().zip(block.senders.iter()) {
```
Comment (mattsse) [inline crates/revm/src/optimism/processor.rs:82]: there's an iterator function for this `BlockWithSenders::transactions_with_sender`

File: crates/primitives/src/block.rs [hunk 4]
```diff
@@ -71,6 +76,17 @@ impl Block {
...
 
+    /// **Expensive**. Transform into a [`BlockWithSenders`] by recovering senders in the contained
+    /// transactions.
+    ///
+    /// Returns `None` if the number of senders does not match the number of transactions in the
+    /// block and the signer recovery for one of the transactions fails.
+    #[track_caller]
+    pub fn with_recovered_senders(self) -> Option<BlockWithSenders> {
+        let senders = self.senders()?;
+        Some(BlockWithSenders { block: self, senders })
```
Comment (mattsse) [inline crates/primitives/src/block.rs:87]: should we return an `Err(Self)` here?

File: crates/primitives/src/block.rs [hunk 5]
```diff
@@ -71,6 +76,17 @@ impl Block {
         BlockWithSenders { block: self, senders }
     }
 
+    /// **Expensive**. Transform into a [`BlockWithSenders`] by recovering senders in the contained
+    /// transactions.
+    ///
+    /// Returns `None` if the number of senders does not match the number of transactions in the
+    /// block and the signer recovery for one of the transactions fails.
+    #[track_caller]
```
Comment (mattsse) [inline crates/primitives/src/block.rs:84]:
```suggestion
```

File: crates/primitives/src/block.rs [hunk 6]
```diff
@@ -71,6 +76,17 @@ impl Block {
         BlockWithSenders { block: self, senders }
     }
 
+    /// **Expensive**. Transform into a [`BlockWithSenders`] by recovering senders in the contained
+    /// transactions.
+    ///
+    /// Returns `None` if the number of senders does not match the number of transactions in the
+    /// block and the signer recovery for one of the transactions fails.
```
Comment (mattsse) [inline crates/primitives/src/block.rs:83]: this returns none if a transaction is invalid

File: bin/reth/src/debug_cmd/in_memory_merkle.rs (hunk 1 above)
Comment (mattsse) [reply bin/reth/src/debug_cmd/in_memory_merkle.rs]: this should map to `BlockValidationError::SenderRecoveryError`

**Review by mattsse — CHANGES_REQUESTED** [review]
should we remove the additional `;`? :)

File: crates/primitives/src/block.rs (hunk 4 above)
Comment (onbjerg) [reply crates/primitives/src/block.rs:87]: we don't use that anywhere so doesn't make sense imo

**Review by mattsse — APPROVED** [review]
lgtm
this forces the caller to do sender recovery and simplifies the functions

**Pull request merged**